src/commands/run.rs

```diff
@@ -454,3 +492,55 @@ fn populate_with_wasi(
 
     Ok(())
 }
+
+#[cfg(not(unix))]
+fn ctx_set_listenfd(num_fd: usize, builder: WasiCtxBuilder) -> Result<(usize, WasiCtxBuilder)> {
+    use listenfd::ListenFd;
+
+    let mut builder = builder;
+    let mut num_fd = num_fd;
+
+    let mut listenfd = ListenFd::from_env();
+
+    for i in 0..listenfd.len() {
+        if let Some(stdlistener) = listenfd.take_tcp_listener(i)? {
+            let _ = stdlistener.set_nonblocking(true)?;
+            let listener = TcpListener::from_std(stdlistener);
+            builder = builder.preopened_listener((3 + i) as _, listener)?;
+            num_fd = 3 + i;
+        }
+    }
+
+    if num_fd > 3 {
+        builder = builder.env("LISTEN_FDS", &(num_fd - 3).to_string())?;
+    }
+
+    Ok((num_fd, builder))
+}
+
+#[cfg(unix)]
+fn ctx_set_listenfd(num_fd: usize, builder: WasiCtxBuilder) -> Result<(usize, WasiCtxBuilder)> {
+    use listenfd::ListenFd;
+
+    let mut builder = builder;
+    let mut num_fd = num_fd;
+
+    for env in ["LISTEN_FDS", "LISTEN_FDNAMES"] {
+        if let Ok(val) = std::env::var(env) {
+            builder = builder.env(env, &val)?;
+        }
+    }
```

Comment:
Wouldn't it make more sense to add a new preopen type returned by `fd_prestat_get` rather than using env vars? Also won't the current code conflict with the fds for pre-opened dirs?

Reply:
My goal was to mimic the listenfd behavior 1:1 so you could reuse already written code.

The preopened dirs are inserted later on, IIRC.. I can recheck again.

Reply:
moved CLI PR to https://github.com/bytecodealliance/wasmtime/pull/3729